dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/task/python/PythonTask.java

```diff
@@ -38,103 +38,110 @@
-  private String buildCommand() throws Exception {
-    String rawPythonScript = pythonParameters.getRawScript().replaceAll("\\r\\n", "\n");
-
-    // replace placeholder
-    Map<String, Property> paramsMap = ParamUtils.convert(ParamUtils.getUserDefParamsMap(taskExecutionContext.getDefinedParams()),
-            taskExecutionContext.getDefinedParams(),
-            pythonParameters.getLocalParametersMap(),
-            CommandType.of(taskExecutionContext.getCmdTypeIfComplement()),
-            taskExecutionContext.getScheduleTime());
-    if (paramsMap != null){
-      rawPythonScript = ParameterUtils.convertParameterPlaceholders(rawPythonScript, ParamUtils.convert(paramsMap));
-    }
+        pythonParameters = JSONUtils.parseObject(taskExecutionContext.getTaskParams(), PythonParameters.class);
 
-    logger.info("raw python script : {}", pythonParameters.getRawScript());
-    logger.info("task dir : {}", taskDir);
-
-    return rawPythonScript;
-  }
+        if (!pythonParameters.checkParameters()) {
+            throw new RuntimeException("python task params is not valid");
+        }
+    }
 
-  @Override
-  public AbstractParameters getParameters() {
-    return pythonParameters;
-  }
+    @Override
+    public void handle() throws Exception {
+        try {
+            //    construct process
+            CommandExecuteResult commandExecuteResult = pythonCommandExecutor.run(buildCommand());
+
+            setExitStatusCode(commandExecuteResult.getExitStatusCode());
+            setAppIds(commandExecuteResult.getAppIds());
+            setProcessId(commandExecuteResult.getProcessId());
+            setVarPool(pythonCommandExecutor.getVarPool());
+        }
+        catch (Exception e) {
+            logger.error("python task failure", e);
+            setExitStatusCode(Constants.EXIT_CODE_FAILURE);
+            throw e;
+        }
+    }
 
+    @Override
+    public void cancelApplication(boolean cancelApplication) throws Exception {
+        // cancel process
+        pythonCommandExecutor.cancelApplication();
+    }
 
+    /**
+     * build command
+     * @return raw python script
+     * @throws Exception exception
+     */
+    private String buildCommand() throws Exception {
+        String rawPythonScript = pythonParameters.getRawScript().replaceAll("\\r\\n", "\n");
+
+        // replace placeholder
+        Map<String, Property> paramsMap = ParamUtils.convert(ParamUtils.getUserDefParamsMap(taskExecutionContext.getDefinedParams()),
+                        taskExecutionContext.getDefinedParams(),
+                        pythonParameters.getLocalParametersMap(),
+                        CommandType.of(taskExecutionContext.getCmdTypeIfComplement()),
+                        taskExecutionContext.getScheduleTime());
+        
+        try {
+            rawPythonScript = VarPoolUtils.convertPythonScriptPlaceholders(rawPythonScript);
+        }
+        catch (StringIndexOutOfBoundsException e) {
+            // TODO: handle exception
```

Comment:
If there is no other way to deal with this error, then you should print the error message